sljit_src/sljitWXExecAllocator.c

```diff
@@ -45,28 +54,53 @@
 	sljit_update_wx_flags((from), (to), (enable_exec))
 
 #ifndef _WIN32
-
+#include <sys/types.h>
 #include <sys/mman.h>
 
+#ifdef __NetBSD__
+#if defined(PROT_MPROTECT)
+#define SLJIT_PROT_WX PROT_MPROTECT(PROT_EXEC)
+#else
+#define SLJIT_PROT_WX 0
+#endif /* PROT_MPROTECT */
+#else /* !NetBSD */
+#define SLJIT_PROT_WX 0
+#endif /* NetBSD */
+
 SLJIT_API_FUNC_ATTRIBUTE void* sljit_malloc_exec(sljit_uw size)
 {
+	static int se_protected = !SLJIT_PROT_WX;
```

Comment:
Is this thread safe?

Reply:
as much as all other static variables used the same way (note that the initial value is defined at compile time) in the codebase.

while not strictly code safe, it will be set to the same value for any concurrent threads anyway, so it is safe

Reply:
Are you sure? This variable is set by `se_protected = mprotect...` and because it fails with one allocation, it may be sucessful with another.

Reply:
the only way mprotect could fail, is if the system has a policy that prevents PROT_EXEC, and that can't change between threads in normal operation.

the binary could be given an exception (which would apply the next time it is loaded)

on the same line, I am assuming a change of system policy (or enforcement) should require a process reload to be effective, but that is not a "race condition"; but if not valid will invalidate the use of this variable to cache that state.

Reply:
I think it's better to proactively check for W^X hardening on a per-platform basis.

Here is a code for this in LLVM:

https://github.com/llvm/llvm-project/blob/524f591a395cbb65a67c85b4df111c5f4e9d753e/compiler-rt/lib/sanitizer_common/sanitizer_linux.cpp#L2186

Whenever CheckMPROTECT() fails, we print an error to stdout and abort. This is verbose to users.

Reply:
the sysctl() function/headers are itself a BSD extension though and so will fail to build in strict posix mode unless "hacked" : 

https://github.com/carenas/sljit/blob/e6e47a4da434a31894f198813f24a7f1266016ef/sljit_src/sljitWXExecAllocator.c#L68-L71

since that is unlikely to be a problem for most users, in supported (or even recently unsupported) versions it might be ok, and it indeed seems preferable to falling back to mprotect()